### DISCUSSION TITLE
LaPlace transform: AttributeError: ‘ImmutableDenseNDimArray’ object has no attribute ‘expand’

### DISCUSSION THREAD

**Discussion by vincentladis** [sympy/sympy#23463] [Q&A] [answered]
Need help to compute the integrate equation of `∫x(t)*e^(-st) dt where x(t`) is my original signal after ifft calculation. The result shows the follwing error AttributeError: `'ImmutableDenseNDimArray' object has no attribute 'expand'` when we convert datas from tensors to numpy array.  While using tensors data as the one after `High pass filter` here the problem `TypeError: unsupported operand type(s) for *: 'Tensor' and 'exp'`. How can I resolve this issue? Thanks in advance.
Here is my code:

`
          import torch
          import numpy as np
          import matplotlib.pyplot as plt
          import os
          from scipy import signal
          import math
          import scipy
          import sympy as sym

      os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"

     f_s = 48000
      f_c = 16000
      f_l = 19000
      f_r = 17000
      b = f_l - f_r
      n_order = 5
      cut_off = 10

     def high_pass_filter():
        # generate a random signal
           array_list = np.random.randn(1, 100, 1000)

    tensor_array = torch.from_numpy(array_list)

    reshape_tensor = tensor_array.reshape(2, 10, -1)

    reshape_img = reshape_tensor.reshape(-1)

    # compute the fft
    fft_signal = torch.fft.fft(reshape_tensor)
    print(fft_signal)
    reshape_fft_signal = fft_signal.reshape(-1)
    normalize_cutoff = (((f_c - b)/cut_off) / 2) / (f_s/2)  # normalize cut_off
    print("normalize : ", normalize_cutoff)
    down = math.sqrt(1 + ((f_c / normalize_cutoff)**(2*n_order)))
    print("down reponse : ", down)

    butterworth = torch.tensor(1 / down)

    G = torch.mul(butterworth, fft_signal)
    G_reshape = G.reshape(-1)
    output_filter = torch.fft.ifft2(G)
    print(output_filter)
    output_filter_reshape = output_filter.reshape(-1)

    plt.figure()
    plt.plot(reshape_img)
    plt.figure()
    plt.plot(reshape_fft_signal)
    plt.figure()
    plt.plot(G_reshape)
    plt.figure()
    plt.plot(output_filter_reshape)
    # plt.show()
    return output_filter

    def pole_zero_plot(poles, zeros, ax=None):
          for pole in poles:
                plt.plot(complex(pole).real, complex(pole).imag, 'rx', markersize=10)
          for zero in zeros:
                 plt.plot(complex(zero).real, complex(zero).imag, 'bo', markersize=10)
  
      if ax is None:
          ax = plt.gca()
      ax.set_aspect(1)

    plt.axis([-2, 2, -2, 2])
    plt.xlabel(r'$\Re\{s\}')
    plt.ylabel(r'$\Im\{s\}')
    plt.grid()

    if __name__ == '__main__':
        sym.init_printing()

    signal = high_pass_filter()
    signal_numpy = signal.detach().numpy()

    t = sym.symbols('t', real=True)
    s = sym.symbols('s', complex=True)

    X = sym.integrate(signal  * sym.exp(-s * t), (t, -sym.oo, sym.oo))
    print(X)

    s0 = sym.symbols('s0', complex=True)
    #
    X, a, cond = sym.laplace_transform(sym.exp(-s0 * t), t, s)
    print(X, ";", a, ";", cond)
    #
    # # Convergence
    #
    sigma, omega = sym.symbols('sigma omega')
    X = 1 / (s + s0)
    X1 = X.subs({s: sigma + sym.I * omega, s0: 1 + sym.I})
    sym.plotting.plot3d(abs(X1), (sigma, -2, 2), (omega, -2, 2),
                         xlabel=r'$\Re\{s\}$', ylabel=r'$\Im\{s\}$', title=r'$|X(s)|$')

    poles = sym.roots(sym.denom(X), s)
    # print(poles)
    #
    zeros = sym.roots(sym.numer(X), s)
    print(zeros)
    #
    X2 = X.subs(s0, 1 + sym.I)
    pole_zero_plot(sym.roots(sym.denom(X2), s), sym.roots(sym.numer(X2), s))

    plt.show()

`

![1](https://user-images.githubusercontent.com/25198529/166635124-d389b4d8-b790-4efd-acac-58876463727f.png)

![2](https://user-images.githubusercontent.com/25198529/166636078-f561ac7b-f460-414d-9a05-9a425e6fd4cf.png)

**Comment by oscarbenjamin** [comment]
I don't think you have copied your code in correctly.

In any case if `signal` is an array of numbers then what do you expect this to do:
```python
X = sym.integrate(signal  * sym.exp(-s * t), (t, -sym.oo, sym.oo))
```

**Reply by vincentladis** [reply]
In the proper way, this line uses `DiracDelta` to multiply with the equation. In my case I switched `DiracDelta` to a generated signal which is the `signal or signal_numpy` in tensor or numpy shape, but raises errors. I want to calculate the impulse response of the signal via the transfer function of the temporal convolution of LAPlace transform  from the latter. Is it possible to proceed through the steps in my script? If so, what type of signal do I need, simple equation or real data? I need you help, thanks

**Reply by oscarbenjamin** [reply, marked as answer]
I don't understand what you're trying to do but my guess is that you should be calculating this numerically rather than symbolically.

**Comment by hanspi42** [comment]
This discussion has fallen asleep, but @oscarbenjamin was totally correct.
